### PR TITLE
Use string.Create instead of concatenation in PdhHelper

### PR DESCRIPTION
**Pull request by xtqqczze**
# PR Summary

* Use string.Create instead of concatenation in method `ConnectToDataSource`
* Use for loop instead of foreach to avoid allocating an enumerator

## PR Context

https://github.com/PowerShell/PowerShell/pull/13410#discussion_r469411428

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [NA] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [NA] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [NA] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [NA] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [NA] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [NA] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by rjmholt** [comment]
I was skeptical of the enumerator for arrays allocating, so I ran a profile:

```csharp
    [MemoryDiagnoser]
    public class LoopAllocationTest
    {
        public static void RunProfile()
        {
            BenchmarkRunner.Run<LoopAllocationTest>();
        }

        private byte[] _nums;

        [GlobalSetup]
        public void Setup()
        {
            _nums = new byte[1000000];
            new Random().NextBytes(_nums);
        }

        [Benchmark(Baseline = true)]
        public void RunForLoop()
        {
            for (int i = 0; i < _nums.Length; i++)
            {
                int k = _nums[i];
            }
        }

        [Benchmark]
        public void RunForEachLoop()
        {
            foreach (int k in _nums)
            {
                ;
            }
        }
    }
```

Results:

```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.450 (2004/?/20H1)
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=5.0.100-preview.4.20258.7
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.25106, CoreFX 5.0.20.25106), X64 RyuJIT
  DefaultJob : .NET Core 5.0.0 (CoreCLR 5.0.20.25106, CoreFX 5.0.20.25106), X64 RyuJIT


|         Method |     Mean |   Error |  StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|--------:|--------:|------:|--------:|------:|------:|------:|----------:|
|     RunForLoop | 228.3 us | 1.41 us | 1.25 us |  1.00 |    0.00 |     - |     - |     - |         - |
| RunForEachLoop | 457.6 us | 4.10 us | 3.43 us |  2.00 |    0.02 |     - |     - |     - |         - |
```

So while it looks like a `for` loop is faster, neither loop allocates.

File: src/Microsoft.PowerShell.Commands.Diagnostics/PdhHelper.cs [hunk 1]
```diff
@@ -542,15 +542,16 @@ public uint ConnectToDataSource(StringCollection blgFileNames)
...
-                doubleNullTerminated += fileName + '\0';
+                string fileName = blgFileNames[i];
+                doubleNullTerminated.Append(fileName).Append('\0');
             }
 
-            doubleNullTerminated += '\0';
+            doubleNullTerminated.Append('\0');
 
-            return ConnectToDataSource(doubleNullTerminated);
+            return ConnectToDataSource(doubleNullTerminated.ToString());
```
Comment (rjmholt) [inline src/Microsoft.PowerShell.Commands.Diagnostics/PdhHelper.cs:554]: I would assume the string allocation is a much bigger issue than the loop enumerator. In which case:

```csharp
var dataString = new StringBuilder();
foreach (string fileName in blgFileNames)
{
    dataString.Append(fileName).Append('\0');
}
dataString.Append('\0');

return ConnectToDataSource(dataString);
```

**Comment by rjmholt** [comment]
Sorry, just realised I for some reason assumed the backing data structure was an array.

Here's an updated benchmark:

```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System;
using System.Collections.Generic;
using System.Collections.Specialized;

namespace CSharpPlayground
{
    [MemoryDiagnoser]
    public class LoopAllocationTest
    {
        public static void RunProfile()
        {
            BenchmarkRunner.Run<LoopAllocationTest>();
        }

        private StringCollection _stringCollection;
        private List<string> _stringList;

        [GlobalSetup]
        public void Setup()
        {
            _stringCollection = new StringCollection();
            _stringList = new List<string>();
            var random = new Random();
            for (int i = 0; i < 1000000; i++)
            {
                string s = random.NextDouble().ToString();
                _stringCollection.Add(s);
                _stringList.Add(s);
            }
        }

        [Benchmark(Baseline = true)]
        public void Run_For_StrCollection()
        {
            for (int i = 0; i < _stringCollection.Count; i++)
            {
                string s = _stringCollection[i];
            }
        }

        [Benchmark]
        public void Run_ForEach_StrCollection()
        {
            foreach (string s in _stringCollection)
            {
                ;
            }
        }

        [Benchmark]
        public void Run_For_List()
        {
            for (int i = 0; i < _stringList.Count; i++)
            {
                string s = _stringList[i];
            }
        }

        [Benchmark]
        public void Run_ForEach_List()
        {
            foreach (string s in _stringList)
            {
                ;
            }
        }
    }
}

```

Results:
```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.450 (2004/?/20H1)
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=5.0.100-preview.4.20258.7
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.25106, CoreFX 5.0.20.25106), X64 RyuJIT
  DefaultJob : .NET Core 5.0.0 (CoreCLR 5.0.20.25106, CoreFX 5.0.20.25106), X64 RyuJIT


|                    Method |       Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------- |-----------:|----------:|----------:|------:|--------:|------:|------:|------:|----------:|
|     Run_For_StrCollection | 4,057.7 us |  80.07 us | 135.97 us |  1.00 |    0.00 |     - |     - |     - |         - |
| Run_ForEach_StrCollection | 5,428.4 us | 105.20 us |  93.25 us |  1.32 |    0.04 |     - |     - |     - |      80 B |
|              Run_For_List |   699.2 us |  13.45 us |  11.23 us |  0.17 |    0.01 |     - |     - |     - |         - |
|          Run_ForEach_List | 3,019.0 us |  35.55 us |  33.25 us |  0.73 |    0.02 |     - |     - |     - |         - |
```

So it seems to be worth switching to `List<string>` and perhaps also keeping the `for` loop for such a dramatic speedup.

**Comment by iSazonov** [comment]
> So it seems to be worth switching to List<string> 

We can not change public API.

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
And it is not hot path - I don't see a value from the PR.

**Comment by rjmholt** [comment]
> We can not change public API.

It's on an internal class, no? If it's exposed publicly, then best not to change it. But this being on a native helper class, it feels like we can make that change.

**Comment by xtqqczze** [comment]
> 
> 
> And it is not hot path - I don't see a value from the PR.

I had assumed this was in a public API so I ignored the fact there are no references to `ConnectToDataSource(StringCollection blgFileNames)` in our codebase.

File: src/Microsoft.PowerShell.Commands.Diagnostics/PdhHelper.cs (hunk 1 above)
Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Diagnostics/PdhHelper.cs:554]: But how can we avoid the string allocation?

**Comment by xtqqczze** [comment]
> 
> So it seems to be worth switching to `List<string>` and perhaps also keeping the `for` loop for such a dramatic speedup.

@rjmholt Given that this method appears to be unused, the performance I wouldn't think changing to `List<string>` justifies the hypothetical performance increase.

Thanks for your help with benchmarking though, it is good to confirm `StringCollection` performance is much worse then generic `List<T>`.

File: src/Microsoft.PowerShell.Commands.Diagnostics/PdhHelper.cs (hunk 1 above)
Comment (rjmholt) [reply src/Microsoft.PowerShell.Commands.Diagnostics/PdhHelper.cs:554]: As in preventing allocating a string altogether? It seems to depend heavily on the requirements of the input and output.

If the input is a fixed, known size in advance, `string.Create()` might be useful.

The output requires a rather idiosyncratic string according to the [docs](https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhbindinputdatasourcew#parameters), so we are forced to allocate one.

**Comment by xtqqczze** [comment]
I have started PR to replace `StringCollection` with generic List<T> in general.

**Comment by xtqqczze** [comment]
> 
> 
> > We can not change public API.
> 
> It's on an internal class, no? If it's exposed publicly, then best not to change it. But this being on a native helper class, it feels like we can make that change.

@rjmholt Would there be value in a generall to PR to change method visibility from `public` to `internal` on `internal` class to avoid confusion like this?

**Comment by rjmholt** [comment]
> Would there be value in a general PR to change method visibility from public to internal on internal class to avoid confusion like this? I have draft PR with these changes: #13436

Generally I personally prefer `public` methods on `internal` and `private` classes; a class/object has a public API and private methods, `internal` implies a secondary API and muddies thinking around the cohesion and contract around a class. However, in PowerShell there is a consideration that once you have an object, any public member on it is visible to PowerShell (even if the class itself is internal), which drives a lot of the other classes that have `internal` members on everything.

File: src/Microsoft.PowerShell.Commands.Diagnostics/PdhHelper.cs [hunk 2]
```diff
@@ -535,22 +535,30 @@ public uint ConnectToDataSource(string dataSourceName)
...
+
+            for (int i = 0; i < blgFileNames.Count; i++)
             {
-                doubleNullTerminated += fileName + '\0';
+                string fileName = blgFileNames[i];
+                doubleNullTerminated.Append(fileName).Append('\0');
             }
 
-            doubleNullTerminated += '\0';
+            doubleNullTerminated.Append('\0');
```
Comment (rjmholt) [inline src/Microsoft.PowerShell.Commands.Diagnostics/PdhHelper.cs:559]: Given the way we calculate capacity up front here, it's relatively simple to move to `string.Create()`:

```csharp
string.Create(capacity, blgFileNames, (buf, fileNames) =>
{
    int offset = 0;
    for (int i = 0; i < fileNames.Count; i++)
    {
        string currFileName = fileNames[i];
        int currFileNameLength = currFileName.Length;
        Span<char> currSpan = buf.Slice(offset, currFileNameLength);
        currFileName.AsSpan().CopyTo(currSpan);

        offset += currFileNameLength;
        buf[offset] = '\0';
        offset++;
    }

    buf[offset] = '\0';
});
```

There's probably a bug or two in the above, but I think the general idea is there

Comment (rjmholt) [reply src/Microsoft.PowerShell.Commands.Diagnostics/PdhHelper.cs:559]: @xtqqczze let me know if you're interested in trying this out. It's probably not a huge win relatively speaking (compared, for example, to native interop marshalling), but still relevant for an allocation/performance improvement PR

Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Diagnostics/PdhHelper.cs]: @rjmholt We can also refactor the following code doing the same thing: https://github.com/PowerShell/PowerShell/blob/3b83a686a1f1b8811dc0480e2b671a2577333ceb/src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs#L2160-L2188

**Comment by xtqqczze** [comment]
@iSazonov Do you any suggestions for CodeFactor issues here?

**Comment by iSazonov** [comment]
@xtqqczze I like the formatting. Please ignore these CodeFactor warnings.

**Comment by rjmholt** [comment]
Do we have a test for this code path? Want to make sure we would identify a regression

**Comment by xtqqczze** [comment]
There appear to be no references to `ConnectToDataSource(StringCollection blgFileNames)`. No tests either.

**Comment by xtqqczze** [comment]
@rjmholt Maybe we could add tests and refactor double null-terminated string generation code to a `NativeUtils` class. We could then reuse the code in `Microsoft.PowerShell.Commands.Management/commands/management/Service.cs`

**Comment by iSazonov** [comment]
@xtqqczze Please resolve merge conflicts.

**Comment by iSazonov** [comment]
@xtqqczze Please resolve merge conflicts.

**Comment by iSazonov** [comment]
The PdhHelper.cs is not in the compile so we should close the PR.

**Comment by rjmholt** [comment]
> The PdhHelper.cs is not in the compile so we should close the PR.

Assuming this is true, I'll close the PR. Let me know if there's something else we should be doing here.

**Pull request closed without merging**